### PR TITLE
Planner improvements for fast printing

### PR DESCRIPTION
**Pull request by tombrazier**
### Description

Whilst attempting to get the fastest 3D benchy possible on my printer, I have found several planner restrictions that prevent faster prints.

#### MINIMUM_PLANNER_SPEED
When accelerating from a standstill, the planner specifies a start speed of `MINIMUM_PLANNER_SPEED` (which has a default value of 0.05 mm/s) and later ensures that this does not result in a step rate less than `MINIMAL_STEP_RATE` (which is hard coded to 120 steps/s).

Generally, 120 steps/s is chosen. Consequently, the first step in the acceleration ramp doesn't happen until 8.33ms has elapsed. This is a very long time when accelerating fast. For example if accelerating at 500,000 steps/s^2 (a very reasonable acceleration these days), the first step should occur at 2ms. The result is stepper motors stalling.

My solution to this problem is to do away with `MINIMUM_PLANNER_SPEED` entirely and, instead, ensure that the minimum speed is at least the average speed for the first step when accelerating from standstill.

#### Jerk speed
When using `CLASSIC_JERK`, it calculates a start speed based on the jerk for the same case where there is an acceleration from standstill. However this value is then ignored and `MINIMUM_PLANNER_SPEED` is used instead.

So this PR adds a line which makes the initial speed the maximum of the jerk-based speed and the aforementioned calculated minimum speed.

#### BLOCK_BUFFER_SIZE
Marlin requires `BLOCK_BUFFER_SIZE` to be a power of 2. This allows a minor planner optimisation because modulo by a power of 2 is just a bitmask operation. However the optimisation is only a few instruction cycles. There is far more benefit to be obtained from a larger block buffer. Many architectures (AVR particularly!) have enough spare SRAM for a few more entries but not enough SRAM to double the buffer size.

This PR removes the power of 2 restriction.

Between these improvements, I have been able to reduce my speed benchy time from about 16 minutes to about 14 minutes on an 8 bit AVR.

### Requirements

Any printer.

### Benefits

Allows higher acceleration without steppers stalling and a lower incidence of slowdowns during the print caused by buffer exhaustion.

### CODE REVIEW

**Comment by thisiskeithb** [comment]
> Marlin requires `BLOCK_BUFFER_SIZE` to be a power of 2. 
> This PR removes the power of 2 restriction.

This line should be also be removed from the config:

https://github.com/MarlinFirmware/Marlin/blob/d1803ec04434a9b5a9e18c12cc392cee15d8f5c6/Marlin/Configuration_adv.h#L2529

**Comment by tombrazier** [comment]
Good catch. Thanks Keith.

**Comment by DerAndere1** [comment]
>  ensure that the minimum speed is at least the average speed for the first step when accelerating from standstill.

The minimum speed is set to `0.5 * acceleration / steps_per_mm`. 
Out of curiosity: Does this mean that the minimum speed is the theoretical speed that would be achieved after half a step period when the printhead would be accelerating constantly from standstill at the set acceleration limit?

EDIT: I think I understood correctly.

**Comment by tombrazier** [comment]
> Out of curiosity: Does this mean that the minimum speed is the theoretical speed that would be achieved after half a step period when the printhead would be accelerating constantly from standstill at the set acceleration limit?

Yes, that will be one way to describe it. I got there via the standard equations of motion for accelerating from zero speed:

distance = 0.5 * acceleration * time^2

We can solve for the time it takes to go one step if distance is measured in steps and acceleration is converted from mm/s/s to steps/s/s:

1 = 0.5 * (acceleration * steps_per_mm) * time^2

time = sqrt(2 / (acceleration * steps_per_mm))

and for average speed (in steps/mm) over that time, it's number of steps (which is 1) divided by time, i.e.:

steps_per_s = sqrt(0.5 * acceleration * steps_per_mm)

to get speed in mm/s,

speed = sqrt(0.5 * acceleration * steps_per_mm) / steps_per_mm = sqrt(0.5 * acceleration / steps_per_mm)

finally, what is actually being calculated is speed squared, so

speed_sqr = 0.5 * acceleration / steps_per_mm

**Comment by classicrocker883** [comment]
> > Out of curiosity: Does this mean that the minimum speed is the theoretical speed that would be achieved after half a step period when the printhead would be accelerating constantly from standstill at the set acceleration limit?
> 
> Yes, that will be one way to describe it. I got there via the standard equations of motion for accelerating from zero speed:
> 
> distance = 0.5 * acceleration * time^2
> 
> We can solve for the time it takes to go one step if distance is measured in steps and acceleration is converted from mm/s/s to steps/s/s:
> 
> 1 = 0.5 * (acceleration * steps_per_mm) * time^2
> 
> time = sqrt(2 / (acceleration * steps_per_mm))
> 
> and for average speed (in steps/mm) over that time, it's number of steps (which is 1) divided by time, i.e.:
> 
> steps_per_s = sqrt(0.5 * acceleration * steps_per_mm)
> 
> to get speed in mm/s,
> 
> speed = sqrt(0.5 * acceleration * steps_per_mm) / steps_per_mm = sqrt(0.5 * acceleration / steps_per_mm)
> 
> finally, what is actually being calculated is speed squared, so
> 
> speed_sqr = 0.5 * acceleration / steps_per_mm

a very precise explanation, very nice.

so this works all fine with 8-bit chip, how about something like 32-bit? now that Block buffer doesn't need to be a power of 2, and it can benefit for a larger size, but not double? so what would be recommended?

**Comment by tombrazier** [comment]
> so this works all fine with 8-bit chip, how about something like 32-bit?

The benefit of minimum speed change applies equally to all platforms.

Some 32 bit processors will be less affected by the BLOCK_BUFFER_SIZE change because they have tons of RAM anyway and it was already possible to choose 32 or 64. However some 32 bit processors are also pretty limited with RAM and will benefit from this.

> Block buffer doesn't need to be a power of 2, and it can benefit for a larger size, but not double? so what would be recommended?

There is a sanity check preventing it from being more than 64, with an error message saying that there is no additional benefit from going beyond 64. I take it that was based on some experimentation. For my part, I found there was a significant difference between 16 and 24 when printing my speed benchy, but a lot less of a benefit in going beyond 24.

There is also a potential downside to a bigger buffer because it takes more processing power to look further ahead into the future. This depends a lot on the gcode being printed, though. It would have to have a large number of very small segments. If there are any longer segments in the queue (i.e. a few mm generally) the acceleration limits will not make any difference to the maximum achievable speeds for the segments that come before that and the planner will therefore not bother to calculate new speed values for these segments.

In practice, I think the limit of 64 makes sense. Even 32 is probably good enough for almost all scenarios. And if you don't have enough RAM for 32, pick the highest number you can.

There is one remaining factor that has bugged me forever and I might one day do something about it. The buffer is implemented as a queue with a `head` element and a `tail` element. By convention, the queue is empty if `head == tail`. This means that a 16 element buffer, for example, can only ever use 15 elements. So that's 75 or more bytes of RAM just being thrown away.

**Comment by classicrocker883** [comment]
>This means that a 16 element buffer, for example, can only ever use 15 elements. So that's 75 or more bytes of RAM just being thrown away.

if I understand how it works, the factor being overlooked is time. time is always passing so as a buffer is being picked up, it is also being put back down. I like to think of it as a road paver. if you ever see one work, it clears the area infront of it and then lays down everything behind it. and everything laid down is like done - set in stone because time passed. time you cannot get back. but the future is ahead of it, that can be tweaked to be the most efficient. same can kinda be said about internal combustion engine. you can have all the air and gas and spark, but if even the exhaust is restricted you will have bad performance.  
basically an engine is only as good as how much it can exhaust. it cant suck in air if it doesnt have means to push back out.
